File: 0871-keys-and-rooms/0871-keys-and-rooms.java
class Solution {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        // 그래프 탐색
        boolean[] visited = new boolean[rooms.size()];
        visited[0] = true;

        dfs(rooms, rooms.get(0), visited);

        for (boolean isVisited : visited) {
            if (!isVisited) return false;
        }
        return true;
    }

    void dfs(List<List<Integer>> rooms, List<Integer> room, boolean[] visited) {
        for (int key : room) {
            if (!visited[key]) {
                visited[key] = true;
                dfs(rooms, rooms.get(key), visited);
            }
        }
    }

    
}

File: 0871-keys-and-rooms/0871-keys-and-rooms.py
class Solution:
    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        visited = [False] * len(rooms)

        def bfs():
            q = deque()
            q.append(rooms[0])
            visited[0] = True
            while q:
                now = q.popleft()
                
                for num in now:
                    if not visited[num]:
                        visited[num] = True
                        q.append(rooms[num])
        bfs()

        return all(visited)
                
        